README
# Rstox version: 1.11.1 (latest alpha, 2019-07-22)
# R version: 3.5


# The package Rstox contains most of the functionality of the stock assesment utility StoX, which is an open source application fo acoustic-trawl and swept-area survey estimation.


# Install the latest official version
devtools::install_github("Sea2Data/Rstox")

# It is also possible to install the latest development version, through devtools::install_github("Sea2Data/Rstox", ref = "develop")
# Note that this does not guarantee a stable version and should ONLY be used to test out new functionality.


# To do this, uncheck the box "32-bit Files" when selecting components to install.
# If you are re-installing an R that has both 32 and 64 bit, you will need to uninstall R first.

# On Windows systems with adminstrator requirements, it is recommended to install R in C:/users/<user>/documents/R.
# Also if you are using Rstudio, please make sure that you are using the correct R version (in case you have
# multiple versions installed). The R version can be selected in Tools > Global Options.

# Note that 64 bit Java is required to run Rstox

# On Windows, install Java from this webpage: https://www.java.com/en/download/windows-64bit.jsp,
# or follow the instructions found on ftp://ftp.imr.no/StoX/Tutorials/

# On Mac, getting Java and Rstox to communicate can be challenging.
# If you run into problems such as "Unsupported major.minor version ...", try the following:
# Update java, on
# 	http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html
# If this does not work install first the JDK and then the JRE:
# 	http://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html
# 	http://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html
# Rstox sohuld also work with Java 11, presently available only as Development Kit:
# 	https://www.oracle.com/technetwork/java/javase/downloads/jdk11-downloads-5066655.html
# You may want to check that the downloaded version is first in the list by running the following in the Terminal:
# 	/usr/libexec/java_home -V
# 	java -version
# Then run this in the Terminal.app (you will be asked for password, but the password will not show as you type.
# It is possible to type the password in a text editor first and then paste it into the Terminal.):
# 	sudo ln -s $(/usr/libexec/java_home)/jre/lib/server/libjvm.dylib /usr/local/lib
# 	sudo R CMD javareconf
# Open R (close and then open if already open) and install rJava:
# 	utils::install.packages('rJava', type='source')
# 	utils::install.packages('rJava', type="binary")
# If this fails, try installing from source instead using utils::install.packages('rJava', type='source')
# Then the installed Rstox should work.


# Release notes for Rstox_1.11.1:

# 1. Added the option ref=NULL in aggregateBySpeciesCategory(), allowing for no reference file for grouping species into catecories.

# 2. Added the option of merging strata by specifying the parameter 'strata' as a list. The other parameters need to be given for the merged strata, so that if three strata are merged the hours/nmi of these should be summed to obtain similar effort as in the unmerged strata.

# 3. Fixed bug in plotAbundance() when there is only one non-NA level of either 'grp1' or 'grp2'.

# 4. Added the function plotFishStation() (and supporting funcitons) to plot the stations of a StoX project by species, possibly in different colors.

# 5. Changed all serialno to serialnumber according to the change from biotic 1.4 to biotic 3.0 applied in StoX as of 2.7.7.

# 6. Fixed bug in writeTransectsMaxSea() which should now only contain non-negative values with correct N/S and E/W. Also changed the parameter projectName of this and related write funcitons to having default projectName=NULL, allowing for extraction of the project name in the input list x.

# 7. Fixed bug in writeTransectsINFO(), where survey coverage included transport.

# 8. Added support for only specifying cruise number in getNMDdata(), in which case data from all cruises are downloaded to a single project named only by the cruise number.

# 9. Removed the process SweptAreaDensity in the test project, as this does not naturally belong to an acoustic-trawl StoX project.

# 10. Fixed bug in readXMLfiles(), where biotic data were duplicated.

# For historical release notes see https://github.com/Sea2Data/Rstox/blob/alpha/NEWS




NEWS
########## Version 1.3.2 ##########

Removed the following unused functions: getParameters(), removeAssignment(), saveRList(), envCall(), getParametersOfProcess(), getBaselineOutput(), rstox.runRProcessFile().

Changed the function getBaseline to return a list of possibly three elements, 'parameters', which are the input parameters of the baseline, 'output', which is the output data from the baseline, and 'processData', which are the process data used in the baseline, specified in the project.xml file (see ?getBaseline).

Added the function getVar() which checks for the existence of a variable in a data frame, and is used to garantee an error if a variable requested from the the Java library of StoX is missing, indicating a change of name.

Removed the need for calling loadEnv() to get the stored data from the environment of a StoX project. loadEnv() has been extended and is should be used at the beginning of all plotting and report functions.

Changed the name of the following functions to starting with a dot, in order to hide these very internal functions from the package INDEX: initiateBootstrap(), jBoolean(), jDouble(), jInt(), Rstox.init().

Added the function getBaselineParameters(), which reads the projec.xml file directly and gets all input parameter values of the baseline, speeding this action up from a couple of seconds to less than 1/10 of a second.

Removed the functions Rstox.demo(), testModel(), and testModel_SweptArea(), which will be replaced by a tutorial in a later version. In the meanwhile, see examples of each function.

Changed the function imputeByAge() and the dependent function distributeAbundance() to store the indices at which age data are missing, and at which these data are extracted by random sampling of data at the same station, strata, and survey, in that prioritised order. The new method is about 5 times faster.

Added the list bootstrapParameters to the output of bootstrapParallel(), saving the seeds and other parameters. Similarly, the list imputParameters is added to the output from imputeByAge(). 

Added the functions getNMDinfo(), getNMDdata(), and the dependent function downloadXML(), used for downloading information and data from the API of the NMD (Norwegian Marine Data center). These functions are not completely finished, and some errors should be expected.

Added a warning in getProcessDataTableAsDataFrame() if no data are found.

Fixed a bug with encoding in xml-files encoded with UTF-8 on Windows, specifically with the nordic characters æ, ø, å. Added the line rapply(x, function(xx) iconv(xx, "UTF-8", "UTF-8"), how="replace") in downloadXML() to assure UTF-8 encoding.

Added the function as.matrix_full() for use when downloading NMD info or data.

Fixed bug in getBaselineOutput(), where projectName was used instead of baseline in the line processNames <- names(runBaseline(baseline, info=TRUE)). 

Small change in bootstrapParallel(), where base.AbByLength and base.AbByInd are now generated prior to invoking cores. Also changed the output from initiateBootstrap() to being the baseline object. 

Changed openProject() to not open a project that does not exist (previously the function mistakedly created the project skeleton if missing).

Added the missing iteration names to the output bootstrapVariable$AbByInd from imputeByAge().

Fixed bug in imputeByAge() and plotAbundance(), which were looking for the object 'runBootstrap_SweptArea', whereas the correct name is bootstrap_SweptArea.



########## Version 1.3.3 ##########

Changed output from the baseline and project.xml to be properly formatted as data frames with numeric and logical type whereever this is possible. Perviously missing values were given as the character "-" (now NA) and logical values as the strings "true" and "false" (nor TRUE and FALSE). This resulted in somewhat faster imputeByAge().

Added the functions getPlots() and getReports(), which can generate all available plots and reports in one call. The functions passes arguments on to the underlying plotting/report functions (all functions with name starting with plot/report).

Added the variables 'level' (level of processing, currntly one of "boostrap" and "impute"), 'var' (the variable to return/plot, currently "abundance" and "weight"), and 'unit' (the unit of the variable, such as "tonnes", and "thousands", or the abbreviations "t", and "1000", see getPlottingUnit() for valid values), along with the variable 'type' (one of "acoustic" and "sweptarea").

Changed loadEnv() to check for already loaded data before loading the data. This functions should called first in all plotting and report functions. Added the function .load_Rstox() used in loadEnv() to load data with old formatting ("-" for NA and "true" for TRUE).

Changed and simplified plotAbundance().

Added the functions getReportVarName() and abbrMatch(), generating the report variable name used by plotting and reporting functions, and matches a string by abbreviation using the method of "$". Rstox now saves plots and reports using the same naming convension, which is level_var_otherparameters, where otherparameters are grp1 and grp2 in the getPlots() and getReports(), which both use the function reportAbundance().

The function polyArea() was changed to take holes in the polygon definitions into account.



########## Version 1.4 ##########

Removed the following unused functions: getParameters(), removeAssignment(), saveRList(), envCall(), getParametersOfProcess(), getBaselineOutput(), rstox.runRProcessFile(), Rstox.demo(), testModel(), and testModel_SweptArea().

Changed the function getBaseline to return a list of possibly three elements, 'parameters', which are the input parameters of the baseline, 'output', which is the output data from the baseline, and 'processData', which are the process data used in the baseline, specified in the project.xml file (see ?getBaseline).


build.R
##### Create and install the package stox in R: #####

##### Load devtools: #####
library("devtools")
##########

buildRstox <- function(buildDir, pkgName="Rstox", version="1.0", Rversion="3.3.1", pckversion=list(), official=FALSE, check=FALSE, exportDir=NULL) {
	
	########## Functions ##########
	# Function used for writing the README file automatically, including package dependencies, R and Rstox version and release notes:
	writeRstoxREADME <- function(READMEfile, NEWSfile, version, Rversion, betaAlpha, betaAlphaString, imports, official=FALSE){
		# Write Rstox and R version in the first two lines. THIS SHOULD NEVER BE CHANGED, SINCE STOX READS THESE TWO LINES TO CHECK VERSIONS:
		write(paste0("# Rstox version: ", version, " (latest ", betaAlphaString, ", ", format(Sys.time(), "%Y-%m-%d"), ")"), READMEfile)
		write(paste0("# R version: ", Rversion), READMEfile, append=TRUE)
		
		write("", READMEfile, append=TRUE)
		# Package description and installation code:
		write("# The package Rstox contains most of the functionality of the stock assesment utility StoX, which is an open source approach to acoustic and swept area survey calculations. Download Rstox from ftp://ftp.imr.no/StoX/Download/Rstox or install by running the following commands in R:", READMEfile, append=TRUE)
		
		write("", READMEfile, append=TRUE)
		write("# Install the packages that Rstox depends on. Note that this updates all the specified packages to the latest (binary) version:", READMEfile, append=TRUE)
		write(paste0("dep.pck <- c(\"", paste0(imports, collapse="\", \""), "\")"), READMEfile, append=TRUE)
		# WARNING: IT IS CRUSIAL TO ENCLUDE THE repos IN THIS CALL, FOR STOX TO SOURCE THE README FILE PROPERLY (RESULTS IN AN ERROR IF ABSENT) IT SEEMS "R CMD BATCH source(TheReadMeFile)" RETURNS AN ERROR WHEN repos IS NOT SET (2016-12-16):
		write("install.packages(dep.pck, repos=\"http://cran.us.r-project.org\", type=\"binary\")", READMEfile, append=TRUE)
		#write("install.packages(dep.pck, type=\"binary\")", READMEfile, append=TRUE)
		
		write("", READMEfile, append=TRUE)
		write("# Install Rstox:", READMEfile, append=TRUE)
		# Get the version string, the name of the Rstox tar file, the ftp root and, finally, the ftp directory and full path to the Rstox tar file:
		# Changed added to make the package name identical to the name of the GitHub release:
		#versionString <- paste0("Rstox_", version)
		versionString <- paste("Rstox", version, sep="_")
		tarName <- paste0(versionString, ".tar.gz")
		ftpRoot <- "ftp://ftp.imr.no/StoX/Download/Rstox"
		if(betaAlpha==3){
			ftpDir <- file.path(ftpRoot, "Versions", "Alpha", versionString)
		}
		else{
			if(official){
				ftpDir <- ftpRoot
			}
			else{
				ftpDir <- file.path(ftpRoot, "Versions", versionString)
			}
		}
		tarFile <- file.path(ftpDir, tarName)
		# Write the Rstox install command:
		write(paste0("install.packages(\"", tarFile, "\", repos=NULL)"), READMEfile, append=TRUE)
		
		write("", READMEfile, append=TRUE)
		write("# Alternatively, install the latest development version from GitHub.", READMEfile, append=TRUE)
		write(paste0("# Note that this does not guarantee a stable version."), READMEfile, append=TRUE)
		write(paste0("# For official versions of Rstox, refer to the ftp server ", ftpDir, " as described above."), READMEfile, append=TRUE)
		write("# Install from github using the devtools package:", READMEfile, append=TRUE)
		write("# devtools::install_github(\"Sea2Data/Rstox\", ref=\"develop\")", READMEfile, append=TRUE)
		write("", READMEfile, append=TRUE)
		write("# R should be installed as the 64 bit version (and 64 bit version ONLY for Windows 10).", READMEfile, append=TRUE)
		write("# On Windows systems with adminstrator requirements, it is recommended to install R in C:/users/<user>/documents/R.", READMEfile, append=TRUE)
		write("", READMEfile, append=TRUE)
		write("# Note that 64 bit Java is required to run Rstox", READMEfile, append=TRUE)
		write("# On Windows, install Java from this webpage: https://www.java.com/en/download/windows-64bit.jsp,", READMEfile, append=TRUE)
		write("# or follow the instructions found on ftp://ftp.imr.no/StoX/Tutorials/", READMEfile, append=TRUE)
		write("# On Mac, getting Java and Rstox to communicate can be challenging.", READMEfile, append=TRUE)
		write("# If you run into problems such as \"Unsupported major.minor version ...\", try the following:", READMEfile, append=TRUE)
		write("# Update java, on", READMEfile, append=TRUE)
		write("# \thttp://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html", READMEfile, append=TRUE)
		write("# If this does not work install first the JDK and then the JRE:", READMEfile, append=TRUE)
		write("# \thttp://www.oracle.com/technetwork/java/javase/downloads/jdk8-downloads-2133151.html", READMEfile, append=TRUE)
		write("# \thttp://www.oracle.com/technetwork/java/javase/downloads/jre8-downloads-2133155.html", READMEfile, append=TRUE)
		write("# You may want to check that the downloaded version is first in the list by running the following in the Terminal:", READMEfile, append=TRUE)
		write("# \t/usr/libexec/java_home -V", READMEfile, append=TRUE)
		write("# \tjava -version", READMEfile, append=TRUE)
		write("# Then run this in the Terminal.app:", READMEfile, append=TRUE)
		write("# \tsudo ln -s $(/usr/libexec/java_home)/jre/lib/server/libjvm.dylib /usr/local/lib", READMEfile, append=TRUE)
		write("# \tsudo R CMD javareconf", READMEfile, append=TRUE)
		write("# Open R (close and then open if already open) and install rJava:", READMEfile, append=TRUE)
		write("# \tinstall.packages('rJava', type='source')", READMEfile, append=TRUE)
		write("# Then the installed Rstox should work.", READMEfile, append=TRUE)
		
		# Write release notes:
		write("", READMEfile, append=TRUE)
		write("", READMEfile, append=TRUE)
		write("# Release notes:", READMEfile, append=TRUE)
	
		# Read the changes:
		l <- readLines(NEWSfile)
		# Split into vesions:
		atversion <- which(substr(l, 1, 1) == "#")
		# Strip off "#", and extract the version string:
		versionStringInChanges <- l[atversion]
		startversion <- regexpr("Version ", versionStringInChanges)
		startversion <- startversion + attributes(startversion)$match.length
		versionStringInChanges <- substring(versionStringInChanges, startversion)
		endversion <- regexpr(" ", versionStringInChanges) - 1
		versionStringInChanges <- substr(versionStringInChanges, 1, endversion)
		# Split into versions and format to insert into the README file:
		l <- split(l, findInterval(seq_along(l), c(atversion, length(l)+1)))
		names(l) <- versionStringInChanges
		# Remove the version line:
		l <- lapply(l, function(xx) xx[substr(xx, 1, 1) != "#"])
		thisl <- l[[version]]
		hasText <- which(nchar(thisl)>1)
		thisl[hasText] <- paste0("# ", seq_along(hasText), ". ", thisl[hasText])
		write(thisl, READMEfile, append=TRUE)
	}
	
	# Functions used for extracting the imports in Rstox, in order to inform about these in the README file. This will not be needed once the package is on CRAN:
	discardBasePackages <- function(x){
		inst <- installed.packages()
		Base <- inst[, "Package"][inst[,"Priority"] %in% c("base", "recommended")]
		sort(setdiff(x, Base))
	}
	getImports <- function(buildDir, version=list()){
		# Read the NAMESPACE file and get the package dependencies:
		buildDirList <- list.files(buildDir, recursive=TRUE, full.names=TRUE)
		imports <- NULL
		if(length(buildDirList)){
			print(buildDirList[basename(buildDirList) == "NAMESPACE"])
			NAMESPACE <- readLines(buildDirList[basename(buildDirList) == "NAMESPACE"])
			atImports <- grep("import", NAMESPACE)
			imports <- NAMESPACE[atImports]
			imports <- sapply(strsplit(imports, "(", fixed=TRUE), "[", 2)
			imports <- sapply(strsplit(imports, ")", fixed=TRUE), "[", 1)
			imports <- unique(sapply(strsplit(imports, ",", fixed=TRUE), "[", 1))
			#inst <- installed.packages()
			#Base <- inst[, "Package"][inst[,"Priority"] %in% c("base", "recommended")]
			#imports <- sort(setdiff(imports, Base))
			imports <- discardBasePackages(imports)
			#notBase <- inst[, "Package"][!(inst[,"Priority"]) %in% "base"]
			#imports <- sort(imports[!imports %in% notBase])
		}
		
		# Add required version for packages:
		if(length(version)){
			# Remove version information for packages that are not present in the imports:
			if(!all(names(version) %in% imports)){
				warning("Not all package versions specified in 'version' are present as imports.")
			}
			version <- version[names(version) %in% imports]
			atversion <- match(names(version), imports)
			for(i in seq_along(version)){
				imports[atversion[i]] <- paste0(imports[atversion[i]], " (>= ", version[i], ")")
			}
		}
		return(imports)
	}
	########## End of functions ##########
	
	
	# Clear the installed package:
	try(lapply(.libPaths(), function(xx) remove.packages(pkgName, xx)), silent=TRUE)
	
	if(length(exportDir)==0){
		exportDir <- file.path(dirname(buildDir), "builds")
	}
	if(length(grep(exportDir, buildDir))>0){
		stop("The 'exportDir' cannot be contained in the 'buildDir', since the exports are build from the 'buildDir'")
	}
	#exportDir <- file.path(buildDir, "bundle")
	manDir <- file.path(buildDir, "man")
	DESCRIPTIONfile <- file.path(buildDir, "DESCRIPTION")
	NAMESPACEfile <- file.path(buildDir, "NAMESPACE")
		unlink(NAMESPACEfile, recursive=TRUE, force=TRUE)
	onLoadFile = file.path(buildDir, "R", "onLoad.R")
	onAttachFile = file.path(buildDir, "R", "onAttach.R")
	
	# Changed added to make the package name identical to the name of the GitHub release:
	#thisExportDir <- file.path(exportDir, paste(pkgName, version, sep="_"))
	thisExportDir <- file.path(exportDir, paste(pkgName, version, sep="_"))
	suppressWarnings(dir.create(thisExportDir))
	READMEfile <- file.path(buildDir, "README")
	
	READMEfileExport <- file.path(thisExportDir, "README")
	NEWSfile <- file.path(buildDir, "NEWS")
	
	
	##### Save the following content to the onLoad.R file in the "R" directory: #####
	# JAVA_HOME is unset to be able to load rJava.dll in R CMD BATCH
	# jPackage is moved to rstox.init for dynamic import of rJava
	# The local Rstox environment is created here, in which all useful outputs from functions are placed, and saved at the end of any code:
	onLoadText = paste(
		".onLoad <- function(libname, pkgname){",
		"	",
		"	if(Sys.getenv(\"JAVA_HOME\")!=\"\") Sys.setenv(JAVA_HOME=\"\")",
		"	options(java.parameters=\"-Xmx2g\")",
		"# Create a Rstox environment in which the baseline objects of the various projects are placed. This allows for a check for previously run baseline models and avoids memory leakage:", 
		"	assign(\"RstoxEnv\", new.env(), envir=.GlobalEnv)",
		"	# Assign fundamental variables to the RstoxEnv:",
		"	Definitions <- list(",
		"		StoXFolders = c(\"input\", \"output\", \"process\"), ",
		"		NMD_data_types = c(\"echosounder\", \"biotic\", \"landing\"), ",
		"		StoX_data_types = c(\"acoustic\", \"biotic\", \"landing\"), ",
		"		StoX_data_type_keys = c(acoustic=\"echosounder_dataset\", biotic=\"missions xmlns\", landing=\"Sluttseddel\"), ",
		"		model_types = c(\"AcousticTrawl\", \"SweptAreaLength\", \"SweptAreaTotal\"), ",
		"		processLevels = c(\"bootstrap\", \"bootstrapImpute\")",
		"		)",
		"	assign(\"Definitions\", Definitions, envir=get(\"RstoxEnv\"))",
		"	assign(\"Projects\", list(), envir=get(\"RstoxEnv\"))",
	"}", sep="\n")
	write(onLoadText, onLoadFile)
	##########
	
	##### Save a Java memory message to the onAttach.R file in the "R" directory: #####
	onAttachText = paste(
		".onAttach <- function(libname, pkgname){",
		"	",
		# paste0("	packageStartupMessage(\"", pkgName, "_", version, "\n**********\nIf problems with Java Memory such as java.lang.OutOfMemoryError occurs, try increasing the Java memory by running options(java.parameters=\\\"-Xmx4g\\\"), and possibly using an even higher value than 4g\n**********\n\", appendLF=FALSE)"),
		paste0("	packageStartupMessage(\"", pkgName, "_", version, "\n**********\nIf problems with Java Memory such as java.lang.OutOfMemoryError occurs, try increasing the Java memory by running setJavaMemory(4e9), and possibly using an even higher value than 4 gigabytes\n**********\n\", appendLF=FALSE)"),
	"}", sep="\n")
	write(onAttachText, onAttachFile)
	##########
	
	##### Add required fields to the DESCRIPTION file (below is the full content of the DESCRIPTION file): #####
	# Depends is replaced by @import specified by functions"
	DESCRIPTIONtext = paste(
		paste0("Package: ", pkgName),
		"Title: Running Stox functionality independently in R",
		paste0("Version: ", version),
		"Authors@R: c(",
		"  person(\"Arne Johannes\", \"Holmin\", role = c(\"aut\",\"cre\"), email = \"arnejh@imr.no\"),",
		"  person(\"Edvin\", \"Fuglebakk\", role = \"ctb\"),",
		"  person(\"Gjert Endre\", \"Dingsoer\", role = \"ctb\"),",
		"  person(\"Aasmund\", \"Skaalevik\", role = \"ctb\"),",
		"  person(\"Espen\", \"Johnsen\", role = \"ctb\"))",
		"Author: Arne Johannes Holmin [aut, cre],",
		"  Edvin Fuglebakk [ctr],",
		"  Gjert Endre Dingsoer [ctr],",
		"  Aasmund Skaalevik [ctr],",
		"  Espen Johnsen [ctr]",
		"Maintainer: Arne Johannes Holmin <arnejh@imr.no>",
		paste0("Depends: R (>= ", Rversion, ")"), 
		"Description: This package contains most of the functionality of the StoX software, which is used for assessment of fish and other marine resources based on biotic and acoustic survey and landings data, among other uses. Rstox is intended for further analyses of such data, facilitating iterations over an arbitrary number of parameter values and data sets.",
		"BugReports: https://github.com/Sea2Data/Rstox/issues", 
		"License: LGPL-3",
		"LazyData: true", sep="\n")
	write(DESCRIPTIONtext, DESCRIPTIONfile)
	##########
	
	##### Create documentation: #####
	# Remove current documentation first:
	unlink(manDir, recursive=TRUE, force=TRUE)
	document(buildDir)
	
	# Alter the DESCRIPTION file to contain the imports listed in the NAMESPACE file:
	imports <- getImports(buildDir, version=pckversion)
	#DESCRIPTIONtext <- readLines(DESCRIPTIONfile)
	if(length(imports)){
		cat("Imports:\n		", file=DESCRIPTIONfile, append=TRUE)
		cat(paste(imports, collapse=",\n		"), file=DESCRIPTIONfile, append=TRUE)
		cat("", file=DESCRIPTIONfile, append=TRUE)
	}
	##########
	
	##### Run R cmd check with devtools: #####
	if(check){
		devtools::check(buildDir)
	}
	##########
	
	### Generate the README file: ###
	betaAlpha <- length(gregexpr(".", version, fixed=TRUE)[[1]]) + 1
	betaAlphaString <- c("", "beta", "alpha")[betaAlpha]
	# Read the NAMESPACE file and get the package dependencies. This is needed since we are not on CRAN:
	writeRstoxREADME(READMEfile, NEWSfile, version, Rversion, betaAlpha, betaAlphaString, imports=getImports(buildDir), official=official)
	file.copy(READMEfile, READMEfileExport, overwrite=TRUE)
	##########
	
	##### Create platform independent bundle of source package: #####
	dir.create(thisExportDir, recursive=TRUE)
	pkgFileVer <- build(buildDir, path=thisExportDir)
	# To comply with GitHub, rename to using hyphen (whereas build() hardcodes using "_"):
	versionString <- paste0("Rstox_", version, ".tar.gz")
	pkgFileVerHyphen <- file.path(thisExportDir, versionString)
	file.rename(pkgFileVer, pkgFileVerHyphen)
	
	##### Unload the package: #####
	unload(buildDir)
	##########
	
	##### Install local source package by utils (independent of dev-tools), and check that it loads: #####
	install.packages(pkgFileVerHyphen, repos=NULL, type="source", lib=.libPaths()[1])
	library(Rstox)
	##########
}



# Define the directory of the working copy:
#arnejh
dir <- "~/Documents/Produktivt/Prosjekt/R-packages/Rstox/Rstox"
# aasmunds
#dir <- "C:/Projects/Sea2Data/NewBeam/trunk/beam/StoX/StoX-Main/src/main/resources/stox/system/r"


# Build 1.6.2:
# buildRstox(dir, version="1.6.2", Rversion="3.3.3", official=FALSE, check=FALSE)


# # Build 1.6.3:
# buildRstox(dir, version="1.6.3", Rversion="3.3.3", official=FALSE, check=FALSE, pckversion=list(data.table="1.10.4-3"))


# Build 1.6.4:
# buildRstox(dir, version="1.6.4", Rversion="3.4.2", pckversion=list(data.table="1.10.4-3"), official=FALSE, check=FALSE)


# Build 1.6.5:
#buildRstox(dir, version="1.6.5", Rversion="3.4.2", pckversion=list(data.table="1.10.4-3"), official=FALSE, check=FALSE)


# Build 1.7:
#buildRstox(dir, version="1.7", Rversion="3.4.2", pckversion=list(data.table="1.10.4-3"), official=TRUE, check=FALSE)

# Build 1.7.1:
#buildRstox(dir, version="1.7.1", Rversion="3.4.2", pckversion=list(data.table="1.10.4-3"), official=FALSE, check=FALSE)

# Build 1.7.2:
buildRstox(dir, version="1.7.2", Rversion="3.4.2", pckversion=list(data.table="1.10.4-3"), official=FALSE, check=FALSE)

# Build 1.8:
#buildRstox(dir, version="1.8", Rversion="3.4.3", pckversion=list(data.table="1.10.4-3"), official=TRUE, check=FALSE)
